File: src/nylon/parser/parsers/IfStatementParser.java
package nylon.parser.parsers;

import nylon.NylonException;
import nylon.NylonObject;
import nylon.nylonobjects.NylonFunction;
import nylon.parser.CharIterator;
import nylon.parser.NylonParser;
import nylon.parser.Parser;

import java.util.ArrayList;
import java.util.Objects;
import java.util.Stack;

/**
 * Created by Aedan Smith.
 */

public class IfStatementParser {
    public static final char
            TRUE = '?',
            FALSE = '¿',
            GREATER = '>',
            LESS = '<',
            EQUALS = '=',
            ELSE = '!';

    public static void addTo(ArrayList<Parser> parsers) {
        parsers.set(TRUE, IfStatementParser::parse);
        parsers.set(FALSE, IfStatementParser::parse);
        parsers.set(GREATER, IfStatementParser::parse);
        parsers.set(LESS, IfStatementParser::parse);
        parsers.set(EQUALS, IfStatementParser::parse);
    }

    public static NylonFunction parse(CharIterator in, NylonParser nylonParser) {
        char[] ifs = in.until(stringIterator -> in.hasNext() &&
                (in.peek() == TRUE
                        || in.peek() == FALSE
                        || in.peek() == LESS
                        || in.peek() == GREATER
                        || in.peek() == EQUALS
                        || in.peek() == ELSE
                )
        ).toCharArray();

        NylonFunction ifTrue = nylonParser.parse(in);

        in.skipWhitespace();
        NylonFunction ifFalse = new NylonFunction("EmptyFunction") {
            @Override
            public void applyImpl(Stack<NylonObject> stack) throws NylonException {

            }

            @Override
            public String toString() {
                return "EmptyFunction";
            }
        };
        if (in.hasNext() && in.peek() == ELSE) {
            in.skip();
            ifFalse = nylonParser.parse(in);
        }

        boolean pop = ifs[ifs.length - 1] != ELSE;
        NylonFunction finalIfFalse = ifFalse;
        return new NylonFunction("IfStatement(" + new String(ifs) + ")") {
            @Override
            public void applyImpl(Stack<NylonObject> stack) throws NylonException {
                boolean b = false;
                for (char c : ifs) {
                    switch (c) {
                        case TRUE:
                            b = stack.peek().toBoolean();
                            break;
                        case FALSE:
                            b = !stack.peek().toBoolean();
                            break;
                        case GREATER: {
                            NylonObject o1 = stack.peek(), o2 = stack.get(stack.size() - 2);
                            b = o2.toDouble() > o1.toDouble();
                            break;
                        }
                        case LESS: {
                            NylonObject o1 = stack.peek(), o2 = stack.get(stack.size() - 2);
                            b = o2.toDouble() < o1.toDouble();
                            break;
                        }
                        case EQUALS: {
                            NylonObject o1 = stack.peek(), o2 = stack.get(stack.size() - 2);
                            b = Objects.equals(o1.value, o2.value);
                            break;
                        }
                    }
                    if (b)
                        break;
                }
                if (pop)
                    stack.pop();
                if (b) {
                    ifTrue.apply(stack);
                } else {
                    finalIfFalse.apply(stack);
                }
            }

            @Override
            public String toString() {
                return "If[" + new String(ifs) + "]Then[" + ifTrue + "]Else[" + finalIfFalse + "]";
            }
        };
    }
}

File: src/nylon/parser/parsers/IfStatementParser.kt
package nylon.parser.parsers

import nylon.NylonFunction
import nylon.NylonStack
import nylon.emptyFunction
import nylon.parser.*
import nylon.parser.CharIterator
import java.util.*

/**
 * Created by Aedan Smith.
 */

class IfStatementParserBuilder : ParserBuilder {
    override fun accept(parsers: ArrayList<Parser>) {
        parsers['?'.toInt()] = IfStatementParser()
        parsers['¿'.toInt()] = IfStatementParser()
        parsers['>'.toInt()] = IfStatementParser()
        parsers['<'.toInt()] = IfStatementParser()
        parsers['='.toInt()] = IfStatementParser()
    }
}

class IfStatementParser : Parser {
    override fun apply(src: CharIterator, parser: NylonParser): NylonFunction? {
        val statement = ArrayList<Char>()
        while (src.peek() == '?' || src.peek() == '¿' || src.peek() == '>' || src.peek() == '<' || src.peek() == '=' || src.peek() == '!') {
            statement.add(src.next())
        }
        val ifTrue = parser.parse(src)
        val ifFalse: NylonFunction

        src.skipUntilStatement()
        if (src.hasNext() && src.peek() == '!') {
            src.next()
            ifFalse = parser.parse(src) as NylonFunction
        } else {
            ifFalse = emptyFunction()
        }

        val pop = statement[statement.size - 1] != '!'

        return object : NylonFunction("If(${String(statement)})<$ifTrue>!<$ifFalse>") {
            override fun apply(stack: NylonStack, args: ArrayList<NylonFunction>) {
                var b = false
                for (it in statement) {
                    when (it) {
                        '?' -> b = stack.peek().toBoolean()
                        '¿' -> b = !stack.peek().toBoolean()
                        '<' -> b = stack.peek(2) < stack.peek()
                        '>' -> b = stack.peek(2) > stack.peek()
                        '=' -> b = stack.peek() == stack.peek(2)
                        '!' -> {
                        }
                        else -> throw RuntimeException("Unrecognized if statement '$it'")
                    }
                    if (b) break
                }
                if (pop)
                    stack.pop()
                if (b) {
                    ifTrue!!.apply(stack)
                } else {
                    ifFalse.apply(stack)
                }
            }
        }
    }

    private fun String(statement: ArrayList<Char>): String {
        var string = ""
        statement.forEach { string += it }
        return string
    }
}
